internal/sidecarlogresults/sidecarlogresults.go

```diff
@@ -0,0 +1,111 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sidecarlogresults
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// SidecarLogResult holds fields for storing extracted results
+type SidecarLogResult struct {
+	Name  string
+	Value string
+}
+
+func fileExists(filename string) (bool, error) {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, fmt.Errorf("error checking for file existence %v", err)
+	}
+	return !info.IsDir(), nil
+}
+
+func waitForStepsToFinish(runDir string) error {
+	steps := make(map[string]bool)
+	files, err := ioutil.ReadDir(runDir)
+	if err != nil {
+		return fmt.Errorf("error parsing the run dir  %v", err)
+	}
+	for _, file := range files {
+		steps[filepath.Join(runDir, file.Name(), "out")] = true
+	}
+	for len(steps) > 0 {
+		for stepFile := range steps {
+			// check if there is a post file without error
+			exists, err := fileExists(stepFile)
+			if err != nil {
+				return fmt.Errorf("error checking for out file's existence %v", err)
+			}
+			if exists {
+				delete(steps, stepFile)
+				continue
+			}
+			// check if there is a post file with error
+			if exists, err = fileExists(fmt.Sprintf("%s.err", stepFile)); exists || err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// LookForResults waits for results to be written out by the steps
+// in their results path and prints them in a structured way to its
+// stdout so that the reconciler can parse those logs.
+func LookForResults(runDir string, resultsDir string, resultNames string) error {
+	if err := waitForStepsToFinish(runDir); err != nil {
+		return fmt.Errorf("error while waiting for the steps to finish  %v", err)
+	}
+	expectedResults := []string{}
+	for _, s := range strings.Split(resultNames, ",") {
+		expectedResults = append(expectedResults, s)
+	}
+	g := new(errgroup.Group)
+	for _, resultFile := range expectedResults {
+		resultFile := resultFile
+
+		g.Go(func() error {
+			value, err := ioutil.ReadFile(filepath.Join(resultsDir, resultFile))
+			if os.IsNotExist(err) {
+				return nil
+			} else if err != nil {
+				return fmt.Errorf("error reading the results file %v", err)
+			}
+			newResult := SidecarLogResult{
+				Name:  resultFile,
+				Value: string(value),
+			}
+			res, _ := json.Marshal(&newResult)
+			// print the result to stdout
+			fmt.Println(string(res))
```

Comment:
I'd have the func take in a io.Writer - this will let you test that you're generating the correct output.

Reply:
you could also get fancier and use the writer to encode the output directly - https://go.dev/play/p/NSYd5DrDnd_R

Reply:
Thanks! That indeed allowed me to test the expected output.